ui/fancy/src/js/index.js

```diff
@@ -0,0 +1,157 @@
+require('../css/style.less');
+
+(() => {
+  const GameEngine = require('../../../../lib/engine');
+
+  let currentGame;
+  let mainLoop;
+
+  const blockElements = [];
+
+  const swapper = {
+    x: 0,
+    y: 0,
+
+    getElements(x, y) {
+      const index = x + (y * currentGame.width);
+
+      return [blockElements[index], blockElements[index + 1]];
+    },
+
+    get coordinates() {
+      return { x: this.x, y: this. y };
+    },
+
+    set coordinates(coordinates) {
+      const { x = this.x, y = this.y } = coordinates;
+
+      this.getElements(this.x, this.y).forEach((element) => {
+        element.classList.remove('swapper');
+      });
+      this.getElements(x, y).forEach((element) => {
+        element.classList.add('swapper');
+      });
+      this.x = x;
+      this.y = y;
+    },
+  };
+
+  const keyBindings = {
+    ArrowDown() {
+      if (currentGame && swapper.y < currentGame.height - 1) {
+        swapper.coordinates = { y: swapper.y + 1 };
+      }
+    },
+
+    ArrowUp() {
+      if (currentGame && swapper.y > 0) {
+        swapper.coordinates = { y: swapper.y - 1 };
+      }
+    },
+
+    ArrowLeft() {
+      if (currentGame && swapper.x > 0) {
+        swapper.coordinates = { x: swapper.x - 1 };
+      }
+    },
+
+    ArrowRight() {
+      if (currentGame && swapper.x + 2 < currentGame.width) {
+        swapper.coordinates = { x: swapper.x + 1 };
+      }
+    },
+
+    // Vi key bindings which are far more superior than arrow keys.
+    h() { keyBindings.ArrowLeft(); },
+    j() { keyBindings.ArrowDown(); },
+    k() { keyBindings.ArrowUp(); },
+    l() { keyBindings.ArrowRight(); },
+
+    f() {
+      if (currentGame) {
+        currentGame.addEvent(
+          currentGame.time,
+          'swap',
+          swapper.x + (currentGame.width * swapper.y)
+        );
+      }
+    },
+
+    ' ': () => {
+      if (currentGame) {
+        currentGame.addEvent(currentGame.time, 'addRow');
+      }
+    },
+  };
+
+  function update(state) {
+    state.blocks.forEach((block, index) => {
+      const element = blockElements[index];
+
+      element.classList.remove('red', 'green', 'blue');
+      if (block.color) {
+        element.classList.add(block.color);
+      }
+      if (block.swapTimer === -2 || block.swapTimer === 2) {
```

Comment:
You can't really depend on the swap timer having specific break points like this as it depends on the used ruleset.